### PR TITLE
XWIKI-20477: Upgrade to Solr 9

### PR DESCRIPTION
**Pull request by tmortagne**
# Jira URLs

[XWIKI-20477](https://jira.xwiki.org/browse/XWIKI-20477): Upgrade to Solr 9.4.1 and Lucene 9.8.0
[XWIKI-12955](https://jira.xwiki.org/browse/XWIKI-12955): Upgrade to JAX-RS 2.1.6 (JSR370)
[XCOMMONS-2407](https://jira.xwiki.org/browse/XCOMMONS-2407): Upgrade to Antlr 4.13.1
[XWIKI-21709](https://jira.xwiki.org/browse/XWIKI-21709): Upgrade to Jetty Client 10.0.19
[XWIKI-12674](https://jira.xwiki.org/browse/XWIKI-12674): Replace Restlet with Jersey
[XWIKI-19813](https://jira.xwiki.org/browse/XWIKI-19813): Automatically migrate Solr cores when the major version changes

# Changes

## Description

The main goal was to upgrade to Solr 9.

But upgrading Solr also requires:
* Upgrade to JAX-RS 2.1
* Upgrade to Jetty Client 10.0.19+
* Upgrade to Antlr 4.11.1+

To make migration of some cores from Solr 8 to Solr 9 easier (which means having both versions of the core usable at the same time) the naming of the cores changed a bit: they are now all suffixed with the major version of Solr. 

A new XWikiSolrCore was introduced to hold various information about the core. We used to pass around the SolrClient, but it's missing some information like the core name, and it will be interesting to have a more obvious location to put some Solr core related tool.

Anything that was using Restlet classes (quite a lot more than expected...) moved to using pure JAX-RS APIs (and in some rare cases Servlet API).

### JAX-RS and Jersey

Solr moved to JAX-RS 2. Problem is that Restlet is not compatible with it so we had to find another implementation and since Solr uses Jersey now as JAX-RS implementation, moving to Jersey ourselves is what made the most sense (even if there are other implementations of JAX-RS 2.1).

Restlet had quite a lot of non JAX-RS standard hacks and while I kept some that I have hit during tests (for example forcing a specific media type through the URL parameter `?media=json`) I did not keep everything. Those are mainly located in `org.xwiki.rest.internal.PreMatchingRequestFilter`.

While JAX-RS fixed quite a few Restlet bugs in the implementation of `javax.ws.rs.core.UriBuilder` it's very possible that someone might have relied on one of those bugs. For example, I found code in `org.xwiki.test.ui.TestUtils` which was expecting `/` to not be escaped in elements passed to `UriBuilder#build` (which does not really make any sense).

We are proxying the internal tool used by Jersey to do injection because XWiki JAX-RS components (anything implementing `org.xwiki.rest.XWikiRestComponent`) are expected to be XWiki components too with full support of the XWiki component system (and especially injection of other XWiki component and other XWiki component related tools and metadata). Another consequence is that we had to take care of the injection of JAX-RS `javax.ws.rs.core.Context` annotated fields on XWiki side.

### Antlr 4.10+

Any lexer which was built with Antlr <4.10 won't work anymore (for example, it's the case of the [dokuwiki syntax parser](https://extensions.xwiki.org/xwiki/bin/view/Extension/DokuWiki/DokuWiki%20Syntax%20Parser/)).

# Executed Tests

Not much in terms of new tests, but those changes impact most integration tests (anything which manipulates an XWiki REST endpoint or Solr).

There is still one planned integration test: add an XWiki upgrade test with custom data containing Solr 8 cores.

# Expected merging strategy

* Prefers squash: No
* No backport

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/ImageIT.java [hunk 1]
```diff
@@ -65,7 +65,14 @@
...
  * @since 14.7RC1
  */
-@UITest
+@UITest(
+    extraJARs = {
+        // It's currently not possible to install a JAR contributing a Hibernate mapping file as an Extension. Thus
+        // we need to provide the JAR inside WEB-INF/lib. See https://jira.xwiki.org/browse/XWIKI-8271
+        "org.xwiki.platform:xwiki-platform-notifications-filters-default"
+    },
+    resolveExtraJARs = true
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/ImageIT.java:74]: Not specifically related to your changes, is it?

**Comment by vmassol** [comment]
> Any lexer which was built with Antlr <4.10 won't work anymore (for example, it's the case of the [dokuwiki syntax parser](https://extensions.xwiki.org/xwiki/bin/view/Extension/DokuWiki/DokuWiki%20Syntax%20Parser/)).

So this is a breaking change we need to document in the backward-compat section of the RN, right?

Are there any other breaking changes you know of? (for example does the move to Jersey has impacts that we need to mention in the RN?).

Thanks for your work on this! Looks complex

File: xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java [hunk 2]
```diff
@@ -138,25 +143,34 @@ public abstract class AbstractSolrCoreInitializer implements SolrCoreInitializer
...
 
     @Inject
-    protected SolrSchemaUtils solrSchemaUtils;
+    protected SolrUtils solrUtils;
 
-    protected SolrClient client;
+    protected XWikiSolrCore core;
 
-    protected Map<String, FieldTypeRepresentation> types;
+    @Deprecated(since = "16.1.0RC1")
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java:156]: Missing some doc to explain what to use now it's deprecated: I saw in one of the initializer that the call should now be `this.core.getClient()` right?

File: xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/DefaultJAXRSUtils.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    @Inject
+    private Container container;
+
+    @Override
+    public Charset getCharset(MediaType mediaType)
+    {
+        if (mediaType != null) {
+            String name = mediaType.getParameters().get(MediaType.CHARSET_PARAMETER);
+
+            if (name != null) {
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/DefaultJAXRSUtils.java:66]: The name can never be empty as ""?

File: xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/JAXRSUtils.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ */
+@Role
+@Unstable
+public interface JAXRSUtils
+{
+    /**
+     * Extract the charset from the media type.
+     * 
+     * @param mediaType the media type of the request
+     * @return the charset of the media type
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/JAXRSUtils.java:49]: missing doc to specify the default returned value when it cannot be guessed from the media type.

File: xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-rest/src/main/java/org/xwiki/notifications/rest/NotificationsResource.java [hunk 5]
```diff
@@ -126,9 +127,9 @@ String getNotificationsRSS(
      * Get notifications matching the given parameters. The POST method is used to allow large content in the
      * parameters. For example, the parameter "blackList" could be very long, and the associated URL with the GET method
      * would be too long (generating HTTP 414 error).
-     *
-     * Note: in the interface, we do not list the parameters, because Restlet does not support @FormParam...
-     * See: https://github.com/restlet/restlet-framework-java/issues/1120
+     * <p>
+     * While the signature of the method changed from Java point of view in 16.0.0, it's still exactly the same API from
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-rest/src/main/java/org/xwiki/notifications/rest/NotificationsResource.java:131]:
```suggestion
     * While the signature of the method changed from Java point of view in 16.1.0, it's still exactly the same API from
```

File: xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java [hunk 6]
```diff
@@ -342,6 +342,129 @@ protected void initializeCoreSchema() throws SolrException
...
+            } catch (Exception e) {
+                throw new SolrException("Failed to save the document", e);
+            }
+
+            // Add it to the list of documents to delete
+            delete(source, sourceCore);
+        }
+
+        // Delete the documents
+        delete(source, targetCore);
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java:440]: Why are the documents deleted from the target core?

File: xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java [hunk 7]
```diff
@@ -138,25 +143,34 @@ public abstract class AbstractSolrCoreInitializer implements SolrCoreInitializer
 
     private static final String SOLR_VERSIONFIELDTYPE_VALUE = "defVal";
 
+    private static final int MIGRATION_BATCH_ROWS = 100;
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java:146]: This seems quite low, couldn't we set this to a higher value like 1000? Or do we really have Solr cores with large documents?

File: xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java [hunk 8]
```diff
@@ -342,6 +342,129 @@ protected void initializeCoreSchema() throws SolrException
...
+                throw new SolrException("Failed to save the document", e);
+            }
+
+            // Add it to the list of documents to delete
+            delete(source, sourceCore);
+        }
+
+        // Delete the documents
+        delete(source, targetCore);
+        commit(sourceCore);
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java:441]: Why is the source but not the target committed? Wouldn't it make sense to first commit the target and then the source to ensure no data is lost?

File: xwiki-platform-distribution/xwiki-platform-distribution-debian/pom.xml [hunk 9]
```diff
@@ -63,7 +63,8 @@
     <debian.homepage>http://www.xwiki.org</debian.homepage>
     <debian.controlDir>${project.build.directory}/maven-deb-control-directory</debian.controlDir>
 
-    <xwikiDataDir>/var/lib/xwiki/data</xwikiDataDir>
+    <debian.xwiki.var>/var/lib/xwiki</debian.xwiki.var>
+    <xwikiDataDir>${debian.xwiki.var}/data</xwikiDataDir>
```
Comment (michitux) [inline xwiki-platform-distribution/xwiki-platform-distribution-debian/pom.xml:67]: Is this really related to this PR?

**Comment by michitux** [comment]
Given my comments on the Solr core migration, is there any test that implicitly or explicitly tests Solr core migration? I didn't see any, but maybe I missed it.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/ImageIT.java (hunk 1 above)
Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/ImageIT.java:74]: No, it's just that without it the generated instance is unusable so I tough I would fix that.

**Comment by tmortagne** [comment]
> Are there any other breaking changes you know of? (for example does the move to Jersey has impacts that we need to mention in the RN?).

Yes, I mentioned stuff that Restlet was adding (like some other special URL parameters) which I did not keep because it was not documented or used anywhere in XWiki.

**Comment by tmortagne** [comment]
> Given my comments on the Solr core migration, is there any test that implicitly or explicitly tests Solr core migration? I didn't see any, but maybe I missed it.

Not yet, I plan to add an XWiki upgrade test based on a custom data which contains a Solr 8 with cores from back then.

File: xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/DefaultJAXRSUtils.java (hunk 3 above)
Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/DefaultJAXRSUtils.java:66]: Does not sound like a valid use case to me, in this case you will get the same UnsupportedCharsetException as you would with any other bad charset.

File: xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/JAXRSUtils.java (hunk 4 above)
Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/JAXRSUtils.java:49]: Indeed, better be explicit, as changing the default charset might be an API breakage.

> when it cannot be guessed from the media type

When no charset is provided at all.

File: xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java (hunk 6 above)
Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java:440]: Indeed, this is wrong. I remember testing a migration by hand, but I think I then simplified the code and looks like I forgot to test again.

File: xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java (hunk 7 above)
Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java:146]: Indeed, the cores which are actually going to be migrated in XWiki Standard are not supposed to have very big document, but it might be the case later or in some extension.

To be safe, I'm going to keep 100 by default and allow AbstractSolrCoreInitializer to provide their own values (could even be more than 1000 for ratings for example, since those are supposed to be very small documents).

**Comment by tmortagne** [comment]
> I plan to add an XWiki upgrade test based on a custom data which contains a Solr 8 with cores from back then

Done, migrated from a custom 12.10.11.
